plugins/source/mixpanel/client/client.go

```diff
@@ -0,0 +1,116 @@
+package client
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/cloudquery/cloudquery/plugins/source/mixpanel/internal/mixpanel"
+	"github.com/cloudquery/plugin-sdk/backend"
+	"github.com/cloudquery/plugin-sdk/plugins/source"
+	"github.com/cloudquery/plugin-sdk/schema"
+	"github.com/cloudquery/plugin-sdk/specs"
+	"github.com/rs/zerolog"
+)
+
+type Client struct {
+	logger     zerolog.Logger
+	sourceSpec specs.Source
+
+	MPSpec   Spec
+	Services *mixpanel.Client
+	Backend  backend.Backend
+}
+
+func New(logger zerolog.Logger, sourceSpec specs.Source, mpSpec Spec, services *mixpanel.Client, bk backend.Backend) Client {
+	return Client{
+		logger:     logger,
+		sourceSpec: sourceSpec,
+		MPSpec:     mpSpec,
+		Services:   services,
+		Backend:    bk,
+	}
+}
+
+func (c *Client) Logger() *zerolog.Logger {
+	return &c.logger
+}
+
+func (c *Client) ID() string {
+	return c.sourceSpec.Name
+}
+
+func Configure(ctx context.Context, logger zerolog.Logger, s specs.Source, o source.Options) (schema.ClientMeta, error) {
+	mpSpec := &Spec{}
+	if err := s.UnmarshalSpec(mpSpec); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal mixpanel spec: %w", err)
+	}
+
+	services, err := getServiceClient(logger, mpSpec)
+	if err != nil {
+		return nil, err
+	}
+
+	cl := New(logger, s, *mpSpec, services, o.Backend)
+	return &cl, nil
+}
+
+func getServiceClient(logger zerolog.Logger, spec *Spec) (*mixpanel.Client, error) {
+	if spec.Secret == "" {
+		// allow empty username for the deprecated project secret method: https://developer.mixpanel.com/reference/project-secret
+		return nil, errors.New("no credentials provided")
+	}
+	if spec.ProjectID < 1 {
+		return nil, errors.New("no project id provided")
+	}
+
+	const dateFormat = "2006-01-02"
+
+	if spec.StartDate == "" {
```

Comment:
Can we move these defaults to a `SetDefaults` method on the Spec object instead of in this function? And probably a `Validate` method as well, which can be called from here. This is fairly standard in our plugins and I think it provides good separation of concerns.

Reply:
Good shout, will do